homeassistant/components/mqtt/notify.py

```diff
@@ -0,0 +1,397 @@
+    ) -> None:
+        """Update the notify service through auto discovery."""
+        config: ConfigType = DISCOVERY_SCHEMA(discovery_payload)
+        # Do not rename a service if that service_name is already in use
+        if (
+            new_service_name := slugify(config[CONF_NAME])
+        ) != self._service_name and _check_notify_service_name(
+            self.hass, config
+        ) is None:
+            return
+        # Only refresh services if service name or targets have changes
+        if (
+            new_service_name != self._service_name
+            or config[CONF_TARGETS] != self._config[CONF_TARGETS]
+        ):
+            services = self.hass.data[MQTT_NOTIFY_SERVICES_SETUP]
+            await self.async_unregister_services()
+            if self._service_name in services:
+                del services[self._service_name]
+            self._config = cast(MqttNotificationConfig, config)
+            self._service_name = new_service_name
+            await self.async_register_services()
+            services[new_service_name] = self
+        else:
+            self._config = cast(MqttNotificationConfig, config)
+        self._commmand_template = MqttCommandTemplate(
+            config.get(CONF_COMMAND_TEMPLATE), hass=self.hass
+        )
+        if CONF_DEVICE in config and self._config_entry:
+            await _update_device(self.hass, self._config_entry, config)
+
+    @property
+    def targets(self) -> dict[str, str]:
+        """Return a dictionary of registered targets."""
+        return {target: target for target in self._config[CONF_TARGETS]}
+
+    async def async_send_message(self, message: str = "", **kwargs):
+        """Build and send a MQTT message."""
+        target = kwargs.get(notify.ATTR_TARGET)
+        if (
+            target is not None
+            and self._config[CONF_TARGETS]
+            and set(target) & set(self._config[CONF_TARGETS]) != set(target)
+        ):
+            _LOGGER.error(
+                "Cannot send %s, target list %s is invalid, valid available targets: %s",
+                message,
+                target,
+                self._config[CONF_TARGETS],
+            )
+            return
+        variables = {
+            "message": message,
+            "name": self._config[CONF_NAME],
+            "service": self._service_name,
+            "target": target or self._config[CONF_TARGETS],
+            "title": kwargs.get(notify.ATTR_TITLE, self._config[CONF_TITLE]),
+        }
+        variables.update(kwargs.get(notify.ATTR_DATA) or {})
+        payload = self._commmand_template.async_render(
+            message,
+            variables=variables,
+        )
+        await mqtt.async_publish(
+            self.hass,
+            self._config[CONF_COMMAND_TOPIC],
+            payload,
+            self._config[CONF_QOS],
+            self._config[CONF_RETAIN],
+            self._config[CONF_ENCODING],
+        )
+
+
+async def _update_device(
```

Comment:
```suggestion
def _update_device(
```

Move this together with the other helpers to make the code easier to read